### DISCUSSION TITLE
D'une simulation à un ensemble de simulations

### DISCUSSION THREAD

**Discussion by pbourdu** [publicodes/publicodes#531] [Ideas]
# Problème

J'évalue l'opportunité d'utiliser `publicodes` dans le cadre de simulation financières. Les possibilités de documentation et d'utilisation du projet sont encourageantes.
Pourtant, assez vite : un use case n'est pour l'instant pas adressé par publicodes. Je m'explique...

L'`Engine` permet de définir une situation, avec des variables en entrée, des résultats en sortie. C'est efficace pour modéliser un domaine compliqué (par exemple la fiscalité 😇).

Lorsqu'on modélise une situation fiscale, il est courant de ne pas s'arrêter à une seule période d'analyse. Par exemple si je dois simuler la fiscalité des revenus fonciers générés par un appartement mis en location nue, la première année m'intéresse tout autant que les 14 (ou plus) années qui suivent.

Un autre cas plus accessible pour imager la problématique technique : le calcul d'un échéancier d'emprunt ([à annuité constante](https://fr.wikipedia.org/wiki/Annuit%C3%A9_constante)).

Ex., 100k emprunté, Taux annuel 3%, Durée 7 ans
Le tableau d'amortissement 
| Période | Capital restant dû | Intérêts | Amortissement | Annuité | Capital restant dû après annuité |
|---------|--------------------|---------|----------|-------------------|----------------------------------|
| 1       | 100000             | 3000    | 13050.64 | 16050.64          | 86949.36                         |
| 2       | 86949.36           | 2608.48 | 13442.16 | 16050.64          | 73507.20                         |
| 3       | 73507.20           | 2205.22 | 13845.42 | 16050.64          | 59661.78                         |
| 4       | 59661.78           | 1789.85 | 14260.79 | 16050.64          | 45400.99                         |
| 5       | 45400.99           | 1362.03 | 14688.61 | 16050.64          | 30712.38                         |
| 6       | 30712.38           | 921.37  | 15129.27 | 16050.64          | 15583.11                         |
| 7       | 15583.11           | 467.49  | 15583.11 | 16050.60          | 0                                |

**`Publicodes` permet de définir pour une période la relation entre les variables et donc de calculer "toute une ligne". Par contre il n'y a pas un niveau de simulation plus haut permettant de définir un ensemble de `situation` (eq. à une ligne de tableau).**

# Idée

Le modèle `Engine` travaille actuellement sur une `situation` (l'équivalent d'une période).

Un modèle `Simulation` pourrait prendre place pour créer et maintenir une suite de `situation` (avec leur `Engine`).

- Une simulation contient des variables définissant les hypothèses (ex. le montant emprunté, le taux d'emprunt, la durée)
- Une simulation a une méthode `calculate` qui : 
  - Génère une `situation` par période (`new Engine`), initialise les variables (`#setSituation`), demande l'évaluation des résultats
  - Respecte l'ordre des `situation`, et alimente les variables nécessaires aux calculs à partir des hypothèses de la simulation **et** des résultats de la `situation` précédente
- Une simulation a une format de restitution des résultats simple, sous forme de tableau

# Questions en suspens
- [ ] Il est possible de réaliser cette approche dans une librairie à part, qui utiliserait `publicodes`. Est-ce que son intégration directement dans le projet peut avoir de la valeur pour la suite ?

**Comment by cdornier** [comment]
Bonjour, 

Effectivement cet ajout serait très intéressant voir indispensable pour les calculs financiers. Nous avons renoncer à utiliser cette lib parce qu'il manquait cette fonctionnalité. 

Clément

**Comment by johangirod** [comment]
J'ai l'impression que le [mécanisme `contexte`](https://publi.codes/docs/mecanismes#contexte) pourrait répondre au cas d'usage.

```yaml
période: 
  par défaut: 1
capital restant dû: 
   valeur: # calcul capital en fonction de la valeur "période"

année 3: 
  avec: 
    capital restant dû : 
      valeur: capital restant dû
      contexte: 
        période: 3
```

Est-ce que cela répondrait à votre besoin ?

Ceci étant dit, cette écriture peut être très verbeuse, car elle nécessite de redéclarer chaque champ pour chaque année. Il faudrait voir si on peut trouver une syntaxe qui simplifie tout cela (notamment en factorisant le contexte entre plusieurs variables).

**Reply by pbourdu** [reply]
Si je comprend bien le mécanisme : il est possible de "déclarer en dur" chaque année. Si je veux avoir un nombre de période dynamique ça me semble ne pas marcher ... 🤔

**Reply by johangirod** [reply]
Oui, tout à fait. Je pense qu'en couplant avec l'ajout de la fonctionnalité liste, il y aurait moyen de répondre à ce cas d'usage. En tout cas, il faut qu'on l'ait en tête lorsqu'on y réfléchira.

#392

En attendant, c'est effectivement une partie à faire hors moteur.

**Reply by pbourdu** [reply]
Intéressant, je vois qu'il y a pas mal de discussions qui touchent de près ou de loin ce topic. Je prendrai le temps de la lecture et peut-être de me connecter à l'un de vos cafés virtuels à l'occasion.

**Comment by johangirod** [comment]
J'ai rédigé une proposition de nouvelle fonctionnalité du langage qui résoudrait le cas d'usage proposé ici : #566 (voir à la fin pour l'implémentation d'un échéancier de prêt). 

Preneur de vos retours @pbourdu : est-ce que la solution proposée vous parait facile à comprendre ? À écrire ?

**Reply by pbourdu** [reply]
> J'ai rédigé une proposition de nouvelle fonctionnalité du langage qui résoudrait le cas d'usage proposé ici : #566 (voir à la fin pour l'implémentation d'un échéancier de prêt). 
> 
> 
> 
> Preneur de vos retours @pbourdu : est-ce que la solution proposée vous parait facile à comprendre ? À écrire ?

Bonjour @johangirod 
Super cette proposition.
Des deux trois cas d'usages que j'ai en tête, ça m'a l'air de faire le job : 
- Calculer un tableau d'amortissement d'emprunt ;
- Gérer plusieurs immeubles dans une declaration de revenus (revenus fonciers)
- ...
Je serai curieux de tester ça quand ça sortira 😛
